Uniform handling of ENCRYPTION_NON_LOCAL

pontusmelke opened the pull request:
Although the old check, which looked for a loopback address was in someway
_more correct_. We want a well-defined uniform across driver. So the driver
will only consider a connection to be local if it matches `localhost`,
`127.x.y.z`, or `::1` (and variants thereof)

pontusmelke committed:
Uniform handling of ENCRYPTION_NON_LOCAL

Although the old check, which looked for a loopback address was in someway
_more correct_. We want a well-defined uniform across driver. So the driver
will only consider a connection to be local if it matches `localhost` or
`127.x.y.z`.

MishaDemianenko commented on driver/src/main/java/org/neo4j/driver/internal/net/BoltServerAddress.java:
```diff
@@ -138,21 +141,12 @@ public int port()
-            // if it's unknown, it's not local so we can safely return false
-            return false;
-        }
+        return LOCALHOST.matcher( host ).matches();
```
Why do we use regexp to know that that address belongs to local instead of InetAddress methods?

thobe replied:
Because this behaviour can be implemented uniformly across all different drivers on all different platforms, such that we have the same behaviour in all drivers.

thobe commented on driver/src/main/java/org/neo4j/driver/internal/net/BoltServerAddress.java:
```diff
@@ -34,6 +34,9 @@
     public static final int DEFAULT_PORT = 7687;
     public static final BoltServerAddress LOCAL_DEFAULT = new BoltServerAddress( "localhost", DEFAULT_PORT );
+    private static final Pattern LOCALHOST =
+            Pattern.compile( "^(localhost|127(\\.\\d+){3})$", Pattern.CASE_INSENSITIVE );
```
What about IPv6 addresses (`::1`)?

technige replied:
Good catch. We have no tests currently testing IPv6. We should have.

pontusmelke replied:
Yes I actually thought about that as well, I'll add some tests.

pontusmelke committed:
Make sure `::1` is a local address

Apart from extending the matcher to also include ipv6 addresses, also added
tests that checks so that we can connect to the database of ipv6.

pontusmelke commented:
@thobe now supports `::1` and variants thereof as well. Added some tests for connecting to the database with ipv6 as well and fixed a bug in how we create `BoltServerAddress` from a `host:port` string that wasn't working for ipv6.

thobe commented on driver/src/main/java/org/neo4j/driver/internal/net/BoltServerAddress.java:
```diff
@@ -34,6 +34,9 @@
     public static final int DEFAULT_PORT = 7687;
     public static final BoltServerAddress LOCAL_DEFAULT = new BoltServerAddress( "localhost", DEFAULT_PORT );
+    private static final Pattern LOCALHOST =
+            Pattern.compile( "^localhost$|^127(?:\\.[0-9]+){0,2}\\.[0-9]+$|^(?:0*\\:)*?:?0*1$", Pattern.CASE_INSENSITIVE );
```
This regex does not look quite right to me. In IPv4, can you really skip the zeroes in the middle?
And do we care about this matching invalid IP addresses?

pontusmelke replied:
I thought `127.0000.0.1` was supported but lets not think about that, I changed it back to a simpler pattern.

technige commented:
Closing as no longer necessary due to #268
